### PR TITLE
gossip: fix race and update locking

### PR DESCRIPTION
**Pull request by BramGruneir**
2 commits:
- gossip: fix race in which we may not trigger a gossip
  When a new gossip comes in, we replace the ready channel with a new one. If we
  don't hold the old ready channel, we may miss the callback if there was a reply
  already in progress.  The solution is to just store the ready channel before
  we release the lock.
- gossip: move lock to anonymous struct
  This brings gossip's server locking in line with our current style.

The outgoing nodeSet seems like it requires the server lock as well.  I'd like to move it into the struct as well.

This is part of work towards #8263.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/8539)

### CODE REVIEW

**Comment by petermattis** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[gossip/server.go, line 146 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-KPDQJAluIM-GbyUmPW-:-KPDQJAluIM-GbyUmPW0:b-wajacb) ([raw file](https://github.com/cockroachdb/cockroach/blob/906d0964f0f5c7e57f5fd4274e089137b667678a/gossip/server.go#L146)):*
> ```Go
> 		// (once the lock is released) and is closed, we still trigger the
> 		// select below.
> 		ready := s.mu.ready
> ```

Good find. This is a subtle and odd locking pattern. In other places in the code we use a channel as a condition variable by doing a broadcast like:

```
select {
  case ready <- struct{}{}:
  default:
}
```

`ready` would have to be buffered to contain a single element: `ready = make(chan struct{}, 1)`. I'm wondering what the current approach achieves that the above doesn't. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-:-KPDRbKuU20zEDPLeunr:bnfp4nl)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[gossip/client.go, line 79 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-KPDUeVW7lLaYkIfCFGb:-KPDUeVW7lLaYkIfCFGc:bu2vjqz) ([raw file](https://github.com/cockroachdb/cockroach/blob/906d0964f0f5c7e57f5fd4274e089137b667678a/gossip/client.go#L79)):*
> ```Go
> 
> // startLocked requires that the gossip server lock is held.
> func (c *client) startLocked(g *Gossip, disconnected chan *client, rpcCtx *rpc.Context, stopper *stop.Stopper) {
> ```

see comment below: i think this extraction should be reverted.

---

*[gossip/gossip.go, line 1047 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-KPDUJAJxpYrkDqPO3dY:-KPDUJAJxpYrkDqPO3dZ:b-h209b0) ([raw file](https://github.com/cockroachdb/cockroach/blob/906d0964f0f5c7e57f5fd4274e089137b667678a/gossip/gossip.go#L1047)):*
> ```Go
> func (g *Gossip) startClient(addr net.Addr) {
> 	c := newClient(addr, g.serverMetrics)
> 	g.clientsMu.Lock()
> ```

this is not the right lock for `startLocked`. I think you should remove startLocked since it is only used (correctly) in one place.

---

*[gossip/server.go, line 146 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-KPDQJAluIM-GbyUmPW-:-KPDTJBogGDOMwZjQzlQ:bbdtdau) ([raw file](https://github.com/cockroachdb/cockroach/blob/906d0964f0f5c7e57f5fd4274e089137b667678a/gossip/server.go#L146)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Good find. This is a subtle and odd locking pattern. In other places in the code we use a channel as a condition variable by doing a broadcast like:
> 
> ```
> select {
>   case ready <- struct{}{}:
>   default:
> }
> ```
> 
> `ready` would have to be buffered to contain a single element: `ready = make(chan struct{}, 1)`. I'm wondering what the current approach achieves that the above doesn't. 
</details>
+1. That also has the marginal benefit of producing less garbage.

Also, can you conceive of a test for this?

---

*[gossip/server.go, line 357 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-KPDVBx7VPdmRqm-48kS:-KPDVBx7VPdmRqm-48kT:b-ck72j0) ([raw file](https://github.com/cockroachdb/cockroach/blob/906d0964f0f5c7e57f5fd4274e089137b667678a/gossip/server.go#L357)):*
> ```Go
> 		s.mu.Unlock()
> 	}
> 	s.mu.Lock()
> ```

this is a lot of lock juggling. just hold it throughout?

---

*[gossip/server.go, line 395 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-KPDV2Z9No57Pvk1Z6s0:-KPDV2Z9No57Pvk1Z6s1:b-6r044v) ([raw file](https://github.com/cockroachdb/cockroach/blob/906d0964f0f5c7e57f5fd4274e089137b667678a/gossip/server.go#L395)):*
> ```Go
> 
> // GetNodeAddr returns the node's address stored in the Infostore.
> func (s *server) GetNodeAddr() util.UnresolvedAddr {
> ```

it's probably more efficient to have this return a `net.Addr`, since the entire `server` struct is already on the heap.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539)*

**Comment by BramGruneir** [comment]
Review status: 0 of 5 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[gossip/client.go, line 79 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-KPDUeVW7lLaYkIfCFGb:-KPE00fPE1QwB02m9LIu:b63it0u) ([raw file](https://github.com/cockroachdb/cockroach/blob/906d0964f0f5c7e57f5fd4274e089137b667678a/gossip/client.go#L79)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> see comment below: i think this extraction should be reverted.
</details>
The only reason the lock was needed was for the nodeID.  I've added it as a function parameter.  Now the status of the gossip lock isn't important.

---

*[gossip/gossip.go, line 1047 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-KPDUJAJxpYrkDqPO3dY:-KPE0DEQrJvmhRxvK7Om:b-3ktfw7) ([raw file](https://github.com/cockroachdb/cockroach/blob/906d0964f0f5c7e57f5fd4274e089137b667678a/gossip/gossip.go#L1047)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this is not the right lock for `startLocked`. I think you should remove startLocked since it is only used (correctly) in one place.
</details>
startClient is always called while under the gossip lock as well.  It's ugly.  Anyway, I removed the requirement of it being locked, see previous reply.

---

*[gossip/server.go, line 146 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-KPDQJAluIM-GbyUmPW-:-KPE0Qp2dPe7BDnTH1-u:b-5wp4pf) ([raw file](https://github.com/cockroachdb/cockroach/blob/906d0964f0f5c7e57f5fd4274e089137b667678a/gossip/server.go#L146)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> +1. That also has the marginal benefit of producing less garbage.
> 
> Also, can you conceive of a test for this?
</details>
Credit has to go to @spencerkimball for finding the root cause.

As for testing, I noticed the issue when trying to fix #8263.  So it will be tested with my fix for that.

I'm going to try to reimplement ready using the newer technique.  I think this was just older code.

---

*[gossip/server.go, line 357 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-KPDVBx7VPdmRqm-48kS:-KPE1927VeFnBmqdG01C:b3bviuj) ([raw file](https://github.com/cockroachdb/cockroach/blob/906d0964f0f5c7e57f5fd4274e089137b667678a/gossip/server.go#L357)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this is a lot of lock juggling. just hold it throughout?
</details>
Sure, I don't see why not. 

---

*[gossip/server.go, line 395 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-KPDV2Z9No57Pvk1Z6s0:-KPE6T8NWYmDqgdHVCgA:b6rednv) ([raw file](https://github.com/cockroachdb/cockroach/blob/906d0964f0f5c7e57f5fd4274e089137b667678a/gossip/server.go#L395)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> it's probably more efficient to have this return a `net.Addr`, since the entire `server` struct is already on the heap.
</details>
I tried to make this change, but it required changing all the locations that rely on this call.  I'm going to leave it as is.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539)*

**Comment by tamird** [comment]
Reviewed 5 of 5 files at r3, 5 of 5 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[gossip/client.go, line 79 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-KPDUeVW7lLaYkIfCFGb:-KPEBP0WcThgnWdBdRTO:b-536de0) ([raw file](https://github.com/cockroachdb/cockroach/blob/906d0964f0f5c7e57f5fd4274e089137b667678a/gossip/client.go#L79)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> The only reason the lock was needed was for the nodeID.  I've added it as a function parameter.  Now the status of the gossip lock isn't important.
</details>
that's fairly ugly, since you're also passing `g` :(

---

*[gossip/server.go, line 395 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-KPDV2Z9No57Pvk1Z6s0:-KPEBh4erQjBy2Ps3DLX:ballxvl) ([raw file](https://github.com/cockroachdb/cockroach/blob/906d0964f0f5c7e57f5fd4274e089137b667678a/gossip/server.go#L395)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> I tried to make this change, but it required changing all the locations that rely on this call.  I'm going to leave it as is.
</details>
All those locations were added in this PR. That doesn't seem like a good reason to leave it as-is.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539)*

**Comment by BramGruneir** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[gossip/client.go, line 79 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-KPDUeVW7lLaYkIfCFGb:-KPEC3ubw28LXXbeN9tW:b-b8ogkn) ([raw file](https://github.com/cockroachdb/cockroach/blob/906d0964f0f5c7e57f5fd4274e089137b667678a/gossip/client.go#L79)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> that's fairly ugly, since you're also passing `g` :(
</details>
Yeah.  The problem is that nodeID is stored in the infostore, which has to be under lock.  We don't seem to ever have more than one infostore, so we could do something similar to how we store the range descriptor in replica to alleviate a lot of this unneeded locking/unlocking.   

TLDR, Gossip seems ripe for a larger refactoring.

---

*[gossip/server.go, line 395 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-KPDV2Z9No57Pvk1Z6s0:-KPEFYoivSPNppzQhW3y:bydly93) ([raw file](https://github.com/cockroachdb/cockroach/blob/906d0964f0f5c7e57f5fd4274e089137b667678a/gossip/server.go#L395)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> All those locations were added in this PR. That doesn't seem like a good reason to leave it as-is.
</details>
sendGossip and receiveGossip require an unresolveaddr.  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539)*

**Comment by petermattis** [comment]
:lgtm: for the `server.ready` part of the change. 

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-:-KPEFn0lo4iL7f3dlGZ0:bsp5p68)*

**Comment by BramGruneir** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[gossip/server.go, line 146 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-KPDQJAluIM-GbyUmPW-:-KPEFtqZyxg9E96GtDHI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/906d0964f0f5c7e57f5fd4274e089137b667678a/gossip/server.go#L146)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> Credit has to go to @spencerkimball for finding the root cause.
> 
> As for testing, I noticed the issue when trying to fix #8263.  So it will be tested with my fix for that.
> 
> I'm going to try to reimplement ready using the newer technique.  I think this was just older code.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539)*

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[gossip/server.go, line 395 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-KPDV2Z9No57Pvk1Z6s0:-KPELOJkgBBl8TgvpsML:b-5duwmb) ([raw file](https://github.com/cockroachdb/cockroach/blob/906d0964f0f5c7e57f5fd4274e089137b667678a/gossip/server.go#L395)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> sendGossip and receiveGossip require an unresolveaddr.  
</details>
You could make this return a reference.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539)*

**Comment by bdarnell** [comment]
LGTM

---

Reviewed 4 of 5 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[gossip/server.go, line 141 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-KPEoNa69l0_4DrjRM9m:-KPEoNa7zxexd-YqRth0:b-yvxx3c) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7131532ae592649b58b75add8d885142db8ee8f/gossip/server.go#L141)):*
> ```Go
> 	for {
> 		s.mu.Lock()
> 		// Store the old ready so that if it gets replaced with a new one
> ```

This comment no longer applies.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-:-KPEoa46UMkRO2Cjquuo:bb6ywmx)*

**Comment by BramGruneir** [comment]
Review status: 1 of 5 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[gossip/server.go, line 395 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-KPDV2Z9No57Pvk1Z6s0:-KPFS4rRuISAZkOIjB6A:brp74ig) ([raw file](https://github.com/cockroachdb/cockroach/blob/906d0964f0f5c7e57f5fd4274e089137b667678a/gossip/server.go#L395)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> You could make this return a reference.
</details>
Sure.  Done.  

---

*[gossip/server.go, line 141 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-KPEoNa69l0_4DrjRM9m:-KPFChxFP7i-yD0mTnq7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c7131532ae592649b58b75add8d885142db8ee8f/gossip/server.go#L141)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This comment no longer applies.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539)*

**Comment by tamird** [comment]
:lgtm: optional nits

---

Reviewed 4 of 4 files at r7.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[gossip/client_test.go, line 189 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-KPFUt3RZ5OKHVYTtLiE:-KPFUt3RZ5OKHVYTtLiF:b5po413) ([raw file](https://github.com/cockroachdb/cockroach/blob/bafb75a0cda29b159d69b5f2f795678b02057af0/gossip/client_test.go#L189)):*
> ```Go
> 	defer stopper.Stop()
> 	local := startGossip(1, stopper, t, metric.NewRegistry())
> 	localAddr := local.GetNodeAddr()
> ```

nit: these can be inlined (that was my thinking when i suggested that this return a pointer)

---

*[gossip/gossip_test.go, line 204 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-KPFV12XMWbADhPX3caF:-KPFV12XMWbADhPX3caG:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/bafb75a0cda29b159d69b5f2f795678b02057af0/gossip/gossip_test.go#L204)):*
> ```Go
> 
> 		for {
> 			localAddr := local.GetNodeAddr()
> ```

ditto

---

*[gossip/gossip_test.go, line 237 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-KPFV2AZ0TJVupR6_RLw:-KPFV2AZ0TJVupR6_RLx:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/bafb75a0cda29b159d69b5f2f795678b02057af0/gossip/gossip_test.go#L237)):*
> ```Go
> 	for i := 0; i < minPeers; i++ {
> 		peer := startGossip(roachpb.NodeID(i+2), stopper, t, metric.NewRegistry())
> 		peerAddr := peer.GetNodeAddr()
> ```

ditto

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-:-KPFUlPGmGbO-YbYLNzn:bw7jlm3)*

**Comment by BramGruneir** [comment]
Review status: 3 of 5 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[gossip/client_test.go, line 189 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-KPFUt3RZ5OKHVYTtLiE:-KPG63vNdODH0qWnGFbh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bafb75a0cda29b159d69b5f2f795678b02057af0/gossip/client_test.go#L189)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: these can be inlined (that was my thinking when i suggested that this return a pointer)
</details>
Done.

---

*[gossip/gossip_test.go, line 204 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-KPFV12XMWbADhPX3caF:-KPG64z9QX1FF8_q1AGB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bafb75a0cda29b159d69b5f2f795678b02057af0/gossip/gossip_test.go#L204)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ditto
</details>
Done.

---

*[gossip/gossip_test.go, line 237 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-KPFV2AZ0TJVupR6_RLw:-KPG65rmO9VncZz6EkqL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bafb75a0cda29b159d69b5f2f795678b02057af0/gossip/gossip_test.go#L237)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ditto
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539)*

**Comment by tamird** [comment]
Reviewed 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[gossip/gossip_test.go, line 204 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-KPFV12XMWbADhPX3caF:-KPG90H8fTSBDcSQxGFC:b-vofic0) ([raw file](https://github.com/cockroachdb/cockroach/blob/bafb75a0cda29b159d69b5f2f795678b02057af0/gossip/gossip_test.go#L204)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> Done.
</details>
i think you missed this one.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539)*

**Comment by BramGruneir** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[gossip/gossip_test.go, line 204 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539#-KPFV12XMWbADhPX3caF:-KPGU03hGRr2zcmvnh76:b-o5brko) ([raw file](https://github.com/cockroachdb/cockroach/blob/bafb75a0cda29b159d69b5f2f795678b02057af0/gossip/gossip_test.go#L204)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> i think you missed this one.
</details>
Ah, sorry, localAddr is used twice in the function.  I audited both test files for all instances of getNodeAddr and moved all appropriate ones.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8539)*
